### PR TITLE
sidecarset support sidecar container hot upgrade

### PR DESCRIPTION
**Pull request by zmberg**
Signed-off-by: liheng.zms <liheng.zms@alibaba-inc.com>
### Ⅰ. Describe what this PR does
sidecarset support sidecar container hot upgrade. 

Refer to the details: https://github.com/zmberg/kruise/blob/sidecarset_hotupgrade_proposal/docs/proposals/20210305-sidecarset-hotupgrade.md

### CODE REVIEW

**Comment by Fei-Guo** [comment]
For changes like this, we need a design doc.

File: pkg/control/sidecarcontrol/util_hotupgrade.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+	}
+	if err := json.Unmarshal([]byte(currentStr), &hotUpgradeWorkContainer); err != nil {
+		klog.Errorf("Parse Pod(%s.%s) annotations(%s) Value(%s) failed: %s", pod.Namespace, pod.Name,
+			SidecarSetWorkingHotUpgradeContainer, currentStr, err.Error())
+		return hotUpgradeWorkContainer
+	}
+	return hotUpgradeWorkContainer
+}
+
+func GetEmptyHotUpgradeContainer(sidecarName string, pod *corev1.Pod) string {
```
Comment (Fei-Guo) [inline pkg/control/sidecarcontrol/util_hotupgrade.go:88]: How about GetPodEmptyHotUpgradeContainer?

File: pkg/controller/sidecarset/sidecarset_hotupgrade.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+		if err := p.Client.Get(context.TODO(), key, podClone); err != nil {
+			klog.Errorf("error getting updated pod(%s.%s) from client", podClone.Namespace, podClone.Name)
+		}
+		return updateErr
+	})
+
+	return err
+}
+
+func flipPodSidecarContainer(control sidecarcontrol.SidecarControl, pod *corev1.Pod) {
```
Comment (Fei-Guo) [inline pkg/controller/sidecarset/sidecarset_hotupgrade.go:69]: How about rename it to "flipPodSidecarContainerDo"?

File: pkg/controller/sidecarset/sidecarset_hotupgrade.go [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+
+	// First, empty hot sidecar container will be upgraded with the latest sidecarSet specification
+	if c1.Image == sidecarContainer.UpgradeStrategy.HotUpgradeEmptyImage {
+		return c1.Name
+	} else if c2.Image == sidecarContainer.UpgradeStrategy.HotUpgradeEmptyImage {
+		return c2.Name
+	}
+
+	// Second, Not ready sidecar container will be upgraded
+	c1Ready, c2Ready := isContainerConsistentAndReady(&c1, pod, control), isContainerConsistentAndReady(&c2, pod, control)
```
Comment (Fei-Guo) [inline pkg/controller/sidecarset/sidecarset_hotupgrade.go:200]: Can I assume that if the code reaches here, it means the previous hot upgrade has not finished yet?  If we continue to update, is it possible to cause "working" sidecar container to be recreated, hence breaking the service?

**Review by Fei-Guo — COMMENTED** [review]
Can use change the sidecar upgradetype from HotUpgrade to ColdUpgrade? If so, what is the expected behavior?

File: apis/apps/v1alpha1/sidecarset_types.go [hunk 4]
```diff
@@ -96,10 +96,20 @@ type SidecarContainerUpgradeType string
...
+	// when sidecar container is stateless, use ColdUpgrade
+	// otherwise HotUpgrade are more HotUpgrade.
+	// examples for istio envoy container is suitable for HotUpgrade
+	// default is ColdUpgrade
 	UpgradeType SidecarContainerUpgradeType `json:"upgradeType,omitempty"`
+
+	// when HotUpgrade, HotUpgradeEmptyImage is used to complete the hot upgrading process
+	// HotUpgradeEmptyImage is consistent of sidecar container in Command, Args, Liveness probe, etc.
+	// but it doesn't work.
+	HotUpgradeEmptyImage string `json:"hotUpgradeEmptyImage,omitempty"`
```
[the remark this answers is not in the document]
Comment (zmberg) [reply apis/apps/v1alpha1/sidecarset_types.go:112]: what means hard-coded image?

File: pkg/controller/sidecarset/sidecarset_hotupgrade.go (hunk 3 above)
Comment (zmberg) [reply pkg/controller/sidecarset/sidecarset_hotupgrade.go:206]: In this scenario, if the last hot upgrade fails (the new version of sidecar does start to fail) and the user adjusts the image and then updates the SidecarSet to trigger the upgrade, the container that is not ready will be selected first to upgrade

Comment (zmberg) [reply pkg/controller/sidecarset/sidecarset_hotupgrade.go:206]: For the scenario you said above I made a judgment in the **IsSidecarSetCanUpgrade** method in the file [sidecarset_control.go](https://github.com/zmberg/kruise/blob/sidecarset_hotupgrade/pkg/control/sidecarcontrol/sidecarset_control.go), for being in the process of upgrading pod this time I will not upgrade.

**Review by Fei-Guo — COMMENTED** [review]
In SidecarControl interface, 

What is the difference between IsPodConsistentAndReady and IsPodUpdatedConsistent? 
The names look like they are duplicated. 

IsPodUpdatedConsistent is not a correct english expression. Maybe use IsPodUpdatedConsistently 

IsSidecarSetCanUpgrade is not a correct english expression. Maybe
IsSidecarSetCanUpgrade ->  IsSidecarSetUpgradable 

There are other naming issues for you to consider, 
IsNeedInjectVolumeMount -> NeedToInjectVolumeMount
NeedInjectOnUpdatedPod -> NeedToInjectInUpdatedPod

File: pkg/controller/sidecarset/sidecarset_hotupgrade.go [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+	klog.V(3).Infof("pod(%s.%s) container(%s) ready(%v) container(%s) ready(%v)", pod.Namespace, pod.Name, c1.Name, c1Ready, c2.Name, c2Ready)
+	if c1Ready && !c2Ready {
+		return c2.Name
+	} else if !c1Ready && c2Ready {
+		return c1.Name
+	}
+
+	// Third, the older sidecar container will be upgraded
+	_, emptyContainer := sidecarcontrol.GetPodHotUpgradeContainers(sidecarContainer.Name, pod)
```
Comment (furykerry) [inline pkg/controller/sidecarset/sidecarset_hotupgrade.go:225]: where is the logic to determine older sidecar?

**Review by furykerry — CHANGES_REQUESTED** [review]

File: pkg/control/sidecarcontrol/util_hotupgrade.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+	// record which hot upgrade container is working currently
+	SidecarSetWorkingHotUpgradeContainer = "kruise.io/sidecarset-working-hotupgrade-container"
+
+	// hotUpgrade container name suffix
+	hotUpgradeNameSuffix1 = "-1"
+	hotUpgradeNameSuffix2 = "-2"
+
+	// sidecar container version in container env(SIDECARSET_VERSION)
+	SidecarSetVersionEnvKey = "SIDECARSET_VERSION"
+	// sidecar container version in container env(SIDECARSET_VERSION_ALT)
```
Comment (furykerry) [inline pkg/control/sidecarcontrol/util_hotupgrade.go:39]: container version env  in the other sidecar container of the same hotupgrade sidecar

**Review by furykerry — CHANGES_REQUESTED** [review]

**Comment by furykerry** [comment]
> NeedToInjectInUpdatedPod
I also find these two methods are somehow overlapped( IsPodConsistentAndReady and IsPodUpdatedConsistent), maybe we can change IsPodConsistentAndReady to IsPodReady, and make an utility function to implement IsPodConsistentAndReady by using IsPodReady+IsPodUpdatedConsistent

File: pkg/control/sidecarcontrol/sidecarset_control.go [hunk 7]
```diff
@@ -143,47 +169,60 @@ func (c *commonControl) IsPodUpdatedConsistent(pod *v1.Pod, ignoreContainers set
...
+	cStatus := map[string]bool{}
+	for _, status := range pod.Status.ContainerStatuses {
+		cStatus[status.Name] = status.Ready
+	}
+	sidecarContainerList := GetSidecarContainersInPod(sidecarSet)
+	for _, sidecar := range sidecarContainerList.List() {
+		// when containerStatus.Ready == true and container non-consistent,
+		// indicates that sidecar container is in the process of being upgraded
+		// wait for the last upgrade to complete before performing this upgrade
+		if cStatus[sidecar] && !c.IsPodUpdatedConsistently(pod, sets.NewString(sidecar)) {
```
Comment (furykerry) [inline pkg/control/sidecarcontrol/sidecarset_control.go:193]: why we have to wait for the last upgrade to complete? can we upgrade even if the last upgrade is in progress?

**Review by furykerry — CHANGES_REQUESTED** [review]

File: pkg/control/sidecarcontrol/sidecarset_control.go (hunk 7 above)
Comment (zmberg) [reply pkg/control/sidecarcontrol/sidecarset_control.go:193]: If we continue to update, it is possible to cause "working" sidecar container to be recreated, hence breaking the service. So waiting for the last upgrade to finish will be a little more business friendly.

> > NeedToInjectInUpdatedPod
> > I also find these two methods are somehow overlapped( IsPodConsistentAndReady and IsPodUpdatedConsistent), maybe we can change IsPodConsistentAndReady to IsPodReady, and make an utility function to implement IsPodConsistentAndReady by using IsPodReady+IsPodUpdatedConsistent

IsPodUpdatedConsistent indicates whether pod.spec and pod.status are consistent after updating the sidecar containers. 
IsPodConsistentAndReady is IsPodUpdatedConsistent + IsPodReady

File: pkg/webhook/pod/mutating/sidecarset.go [hunk 8]
```diff
@@ -27,42 +27,87 @@ import (
...
+	sidecarContainers, sidecarInitContainers, volumesInSidecar, injectedAnnotations, err := buildSidecars(isUpdated, pod, oldPod, injectSidecarSets)
 	if err != nil {
 		return err
 	} else if len(sidecarContainers) == 0 && len(sidecarInitContainers) == 0 {
+		klog.V(3).Infof("[sidecar inject] pod(%s.%s) don't have injected containers", pod.Namespace, pod.Name)
 		return nil
 	}
 
+	// store matched sidecarset list in pod annotations
+	injectedAnnotations[sidecarcontrol.SidecarSetListAnnotation] = sidecarSetNameList
```
[the remark this answers is not in the document]
Comment (zmberg) [reply pkg/webhook/pod/mutating/sidecarset.go:109]: The trigger logic of this place is the pod event, so first we must determine which sidecarSet matches these pods, so I put all the matching logic outside the buildSidecar method

File: pkg/control/sidecarcontrol/sidecarset_control.go (hunk 7 above)
Comment (furykerry) [reply pkg/control/sidecarcontrol/sidecarset_control.go:193]: > If we continue to update, it is possible to cause "working" sidecar container to be recreated, hence breaking the service
it is possible to continue updating by directly upgrade the standby sidecar which is undergo a previous upgrading.

**Review by furykerry — APPROVED** [review]

**Pull request merged**